### ISSUE TITLE
Windows: osquery.db eventually becomes inaccessible & agent disconnects from TLS

### ISSUE THREAD

**Issue by Centurion89** [facebook/osquery#3842]
Osquery Version: 2.8.0
OS: Win10

**TL;DR**: The osquery database on Windows appears to become ~~corrupted~~ inaccessible after 24-72 of uptime. The root cause is currently unknown.

**Problem:** 
While checking in on our TLS server today, I noticed that the vast majority of our agents were showing up as "offline". Further investigation revealed the following issue:
* The osqueryd service is still running
* The osqueryd.exe process is still active
* The TLS server displays the agents as offline
* Stopping the service and manually starting the process results in a database initialization issue:
```
PS > ./osqueryd.exe --flagfile=C:\ProgramData\osquery\osquery.flags --verbose
E1016 14:40:24.601599  6288 init.cpp:601] [Ref #1629] osqueryd initialize failed: Could not initialize database
```

**Troubleshooting steps in progress:**
* Delete the old osquery.db, restart the service 
* Disabled a pack that contains a few odd queries
* Keeping tabs on the few Windows hosts that we recently enrolled that have not yet "crashed"

**Comment by Centurion89** [comment]
I don't think my original assumption about the DB being corrupted is correct. This problem goes away entirely if the host changes ownership of `osquery.db` to be the user. I think this is permissions related.

**Comment by Centurion89** [comment]
Another odd thing (and probably a bug worth triaging) is that when the agent is in this state, starting the osqueryd service doesn't report any problems despite the fact that osquery can't open/write to the database and the service is shown as "running". Stopping the service in this state causes an error:

![image](https://user-images.githubusercontent.com/5092699/31640795-5a8a3064-b295-11e7-8a80-b4fbdf392f10.png)

**Comment by theopolis** [comment]
Is there a zombie osquery process still running on the hosts? That would explain not being able to access the database.

**Comment by Centurion89** [comment]
I think I determined that the "Could not initialize database messages" were actually occurring because I was trying to launch osqueryd from an Administrator Powershell prompt, but the db folder is owned by SYSTEM, so there's no way it would have been able to read it. 

I'll have to re-try testing with a SYSTEM-level prompt tomorrow. I strongly feel that osquery file ownership should be set to "Administrators" to make this type of debugging more straightforward. Dropping into a SYSTEM level shell probably shouldn't be a standard debugging requirement for any tool 😕

**Comment by Centurion89** [comment]
I spoke too soon - I dropped into a SYSTEM prompt and still run into the database initialization error. I have two exact copies of the database:
`osquery.db` - Has all the default permissions, owned by system.
`osquery.db.weakpermissions` - Is owned by Administrators

Using the default `osquery.db` fails:
```
PS C:\ProgramData\osquery> whoami
nt authority\system

PS C:\ProgramData\osquery> .\osqueryd\osqueryd.exe --flagfile=c:\programdata\osquery\osquery.flags --verbose
E1016 22:21:59.216344  4048 init.cpp:601] [Ref #1629] osqueryd initialize failed: Could not initialize database
```
Using `osquery.db.weakpermissions` which is owned by administrators succeeds:
```
PS C:\ProgramData\osquery> .\osqueryd\osqueryd.exe --flagfile=c:\programdata\osquery\osquery.flags --verbose --database_path=c:\programdata\osquery\osquery.db.weakpermissions
```

**Comment by Centurion89** [comment]
Pretty sure @muffins and I tracked this down to an erroneous ACL that was being applied in one of the non-official MSIs for 2.8.0. I wrote some Powershell to patch things up and it seems to work nicely again:

```
c:\ProgramData\osquery\osquery.db 
                                  NT AUTHORITY\SYSTEM:(DENY)(W,D,RD,REA,X,RA)  <<< Should not be there
                                  NT AUTHORITY\SYSTEM:(Rc)
                                  Everyone:(DENY)(W,D,RD,REA,X,RA)
                                  Everyone:(Rc)
                                  NT AUTHORITY\SYSTEM:(I)(OI)(CI)(F)
                                  BUILTIN\Administrators:(I)(OI)(CI)(F)
                                  CREATOR OWNER:(I)(OI)(CI)(IO)(F)
                                  BUILTIN\Users:(I)(OI)(CI)(RX)
                                  BUILTIN\Users:(I)(CI)(WD,AD,WEA,WA)
```

Powershell fix (NOTE: this will change the owner of osquery.db to "administrators"):

```
$path = "c:\programdata\osquery\osquery.db"
TAKEOWN /F $path /A /R /D "Y"
icacls $path /remove:d "NT AUTHORITY\SYSTEM" /T
restart-service osqueryd
```

**Comment by muffins** [comment]
@Centurion89 can you keep us appraised on whether or not that fixes things? Do you think anything will change over time? I only ask as I recall the original issue being something that triggered after a couple days. I'll leave this open so we can keep an eye on things in case there's something more going on than just the DB perms.

**Comment by Centurion89** [comment]
Yep, we should be able to start deploying the fix tomorrow and I'll know within a few days if this starts happening again. I agree, I think it's weird that everything functioned correctly for people even with the ACL present for a few days before they dropped off. Will keep you posted!

**Comment by muffins** [comment]
@Centurion89 any updates on this? Are you still seeing hosts drop off over time?

**Comment by Centurion89** [comment]
We just pushed the fix to the remainder of fleet this morning, so I should definitively know by Wednesday - will post an update then (and hopefully be able to close this one out!) :)

**Comment by Centurion89** [comment]
@muffins Sorry, totally forgot to follow up here. Haven't seen this re-occur since we fixed up permissions. Closing out!
